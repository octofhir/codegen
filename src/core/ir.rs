//! Intermediate Representation (IR) for FHIR types
//!
//! This module defines language-agnostic representations of FHIR resources,
//! datatypes, and primitives that can be transformed into any target language.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Complete type graph representing all FHIR definitions
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TypeGraph {
    /// Resource definitions (e.g., Patient, Observation)
    pub resources: IndexMap<String, ResourceType>,

    /// Complex datatype definitions (e.g., HumanName, Address)
    pub datatypes: IndexMap<String, DataType>,

    /// Primitive type definitions (e.g., string, boolean, date)
    pub primitives: IndexMap<String, PrimitiveType>,

    /// Profile definitions (e.g., USCorePatient)
    pub profiles: IndexMap<String, ProfileType>,

    /// FHIR version this graph represents
    pub fhir_version: FhirVersion,

    /// Metadata about generation
    pub metadata: GraphMetadata,
}

impl TypeGraph {
    /// Create a new empty type graph
    pub fn new(fhir_version: FhirVersion) -> Self {
        Self {
            resources: IndexMap::new(),
            datatypes: IndexMap::new(),
            primitives: IndexMap::new(),
            profiles: IndexMap::new(),
            fhir_version,
            metadata: GraphMetadata::default(),
        }
    }

    /// Add a resource to the graph
    pub fn add_resource(&mut self, name: String, resource: ResourceType) {
        self.resources.insert(name, resource);
    }

    /// Add a datatype to the graph
    pub fn add_datatype(&mut self, name: String, datatype: DataType) {
        self.datatypes.insert(name, datatype);
    }

    /// Add a primitive to the graph
    pub fn add_primitive(&mut self, name: String, primitive: PrimitiveType) {
        self.primitives.insert(name, primitive);
    }

    /// Add a profile to the graph
    pub fn add_profile(&mut self, name: String, profile: ProfileType) {
        self.profiles.insert(name, profile);
    }

    /// Get total number of types in graph
    pub fn total_types(&self) -> usize {
        self.resources.len() + self.datatypes.len() + self.primitives.len() + self.profiles.len()
    }
}

/// Metadata about the type graph
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GraphMetadata {
    /// When this graph was generated
    pub generated_at: String,

    /// Generator version
    pub generator_version: String,

    /// Source packages
    pub source_packages: Vec<String>,

    /// Custom metadata
    pub custom: HashMap<String, serde_json::Value>,
}

impl Default for GraphMetadata {
    fn default() -> Self {
        Self {
            generated_at: chrono::Utc::now().to_rfc3339(),
            generator_version: env!("CARGO_PKG_VERSION").to_string(),
            source_packages: Vec::new(),
            custom: HashMap::new(),
        }
    }
}

/// FHIR version
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum FhirVersion {
    /// FHIR R4 (4.0.x)
    R4,
    /// FHIR R4B (4.3.x)
    R4B,
    /// FHIR R5 (5.0.x)
    R5,
    /// FHIR R6 (6.0.x)
    R6,
}

impl std::fmt::Display for FhirVersion {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FhirVersion::R4 => write!(f, "R4"),
            FhirVersion::R4B => write!(f, "R4B"),
            FhirVersion::R5 => write!(f, "R5"),
            FhirVersion::R6 => write!(f, "R6"),
        }
    }
}

/// FHIR Resource type (e.g., Patient, Observation)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResourceType {
    /// Resource name (e.g., "Patient")
    pub name: String,

    /// Base resource this extends (e.g., Some("DomainResource"))
    pub base: Option<String>,

    /// Properties/elements of this resource
    pub properties: Vec<Property>,

    /// Search parameters defined for this resource
    pub search_parameters: Vec<SearchParameter>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Complex datatype (e.g., HumanName, Address)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DataType {
    /// Datatype name (e.g., "HumanName")
    pub name: String,

    /// Base type this extends (if any)
    pub base: Option<String>,

    /// Properties of this datatype
    pub properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Primitive type (e.g., string, boolean, date)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PrimitiveType {
    /// Primitive name (e.g., "string", "boolean")
    pub name: String,

    /// Base primitive (all primitives extend from Element)
    pub base: Option<String>,

    /// Regex pattern for validation (if any)
    pub pattern: Option<String>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// FHIR Profile (constraint on a resource or datatype)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProfileType {
    /// Profile name (e.g., "USCorePatient")
    pub name: String,

    /// Base type being constrained (e.g., "Patient")
    pub base: String,

    /// Additional constraints on properties
    pub property_constraints: Vec<PropertyConstraint>,

    /// New properties added by this profile
    pub new_properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// Property constraint in a profile
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PropertyConstraint {
    /// Path to the property being constrained
    pub path: String,

    /// New cardinality (if changed)
    pub cardinality: Option<CardinalityRange>,

    /// New type constraints
    pub type_constraints: Vec<String>,

    /// New binding (if changed)
    pub binding: Option<ValueSetBinding>,

    /// Must support flag
    pub must_support: bool,
}

/// A property (field/element) in a resource or datatype
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Property {
    /// Property name (e.g., "name", "birthDate")
    pub name: String,

    /// Original FHIR path (e.g., "Patient.name")
    pub path: String,

    /// Type of this property
    pub property_type: PropertyType,

    /// Cardinality (min..max)
    pub cardinality: CardinalityRange,

    /// Is this a choice element? (e.g., value[x])
    pub is_choice: bool,

    /// If choice, what are the possible types?
    pub choice_types: Vec<String>,

    /// Is this a modifier element?
    pub is_modifier: bool,

    /// Is this a summary element?
    pub is_summary: bool,

    /// Terminology binding (if any)
    pub binding: Option<ValueSetBinding>,

    /// Constraints/invariants
    pub constraints: Vec<InvariantRule>,

    /// Short description
    pub short_description: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Example values
    pub examples: Vec<Example>,
}

/// Type of a property
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum PropertyType {
    /// Primitive type (e.g., "string", "boolean")
    Primitive {
        /// Type name
        type_name: String,
    },

    /// Complex type (e.g., "HumanName", "Address")
    Complex {
        /// Type name
        type_name: String,
    },

    /// Reference to another resource
    Reference {
        /// Allowed target types (empty = any resource)
        target_types: Vec<String>,
    },

    /// Backbone element (inline complex type)
    BackboneElement {
        /// Properties in this backbone element
        properties: Vec<Property>,
    },

    /// Choice type (one of several types)
    Choice {
        /// Possible types for this choice
        types: Vec<String>,
    },
}

impl PropertyType {
    /// Get the primary type name (for simple cases)
    pub fn type_name(&self) -> Option<&str> {
        match self {
            PropertyType::Primitive { type_name } => Some(type_name),
            PropertyType::Complex { type_name } => Some(type_name),
            _ => None,
        }
    }
}

/// Cardinality range (min..max)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct CardinalityRange {
    /// Minimum occurrences
    pub min: u32,

    /// Maximum occurrences (None = unbounded)
    pub max: Option<u32>,
}

impl CardinalityRange {
    /// Create a required single value (1..1)
    pub fn required() -> Self {
        Self { min: 1, max: Some(1) }
    }

    /// Create an optional single value (0..1)
    pub fn optional() -> Self {
        Self { min: 0, max: Some(1) }
    }

    /// Create a required array (1..*)
    pub fn required_array() -> Self {
        Self { min: 1, max: None }
    }

    /// Create an optional array (0..*)
    pub fn optional_array() -> Self {
        Self { min: 0, max: None }
    }

    /// Is this property required?
    pub fn is_required(&self) -> bool {
        self.min >= 1
    }

    /// Is this property an array?
    pub fn is_array(&self) -> bool {
        self.max.is_none() || self.max.is_some_and(|max| max > 1)
    }

    /// Is this property optional?
    pub fn is_optional(&self) -> bool {
        self.min == 0
    }
}

/// ValueSet binding for coded elements
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValueSetBinding {
    /// Binding strength
    pub strength: BindingStrength,

    /// ValueSet canonical URL
    pub value_set: String,

    /// Description
    pub description: Option<String>,
}

/// Binding strength
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum BindingStrength {
    /// Required (must use)
    Required,
    /// Extensible (should use)
    Extensible,
    /// Preferred (could use)
    Preferred,
    /// Example (may use)
    Example,
}

/// Invariant constraint rule
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InvariantRule {
    /// Invariant key (e.g., "pat-1")
    pub key: String,

    /// Severity
    pub severity: ConstraintSeverity,

    /// Human description
    pub human: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// XPath expression (legacy)
    pub xpath: Option<String>,
}

/// Constraint severity
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ConstraintSeverity {
    /// Error (must satisfy)
    Error,
    /// Warning (should satisfy)
    Warning,
}

/// Example value
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Example {
    /// Example label
    pub label: String,

    /// Example value (as JSON)
    pub value: serde_json::Value,
}

/// Documentation for types and properties
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct Documentation {
    /// Short description
    pub short: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Requirements
    pub requirements: Option<String>,

    /// Usage notes
    pub usage_notes: Vec<String>,

    /// Canonical URL
    pub url: Option<String>,
}

/// Search parameter definition
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SearchParameter {
    /// Parameter name/code
    pub code: String,

    /// Parameter type
    pub param_type: SearchParamType,

    /// Description
    pub description: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// Target resource types (for reference params)
    pub target_types: Vec<String>,
}

/// Search parameter type
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SearchParamType {
    /// Number
    Number,
    /// Date/DateTime
    Date,
    /// String
    String,
    /// Token
    Token,
    /// Reference
    Reference,
    /// Composite
    Composite,
    /// Quantity
    Quantity,
    /// URI
    Uri,
    /// Special
    Special,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_type_graph_creation() {
        let graph = TypeGraph::new(FhirVersion::R4);
        assert_eq!(graph.fhir_version, FhirVersion::R4);
        assert_eq!(graph.total_types(), 0);
    }

    #[test]
    fn test_type_graph_add_resource() {
        let mut graph = TypeGraph::new(FhirVersion::R4);

        let resource = ResourceType {
            name: "Patient".to_string(),
            base: Some("DomainResource".to_string()),
            properties: Vec::new(),
            search_parameters: Vec::new(),
            documentation: Documentation::default(),
            url: "http://hl7.org/fhir/StructureDefinition/Patient".to_string(),
            is_abstract: false,
        };

        graph.add_resource("Patient".to_string(), resource);
        assert_eq!(graph.total_types(), 1);
        assert!(graph.resources.contains_key("Patient"));
    }

    #[test]
    fn test_cardinality_required() {
        let card = CardinalityRange::required();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, Some(1));
        assert!(card.is_required());
        assert!(!card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_cardinality_optional() {
        let card = CardinalityRange::optional();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, Some(1));
        assert!(!card.is_required());
        assert!(!card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_array() {
        let card = CardinalityRange::optional_array();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, None);
        assert!(!card.is_required());
        assert!(card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_required_array() {
        let card = CardinalityRange::required_array();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, None);
        assert!(card.is_required());
        assert!(card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_property_type_typename() {
        let prim = PropertyType::Primitive { type_name: "string".to_string() };
        assert_eq!(prim.type_name(), Some("string"));

        let complex = PropertyType::Complex { type_name: "HumanName".to_string() };
        assert_eq!(complex.type_name(), Some("HumanName"));

        let reference = PropertyType::Reference { target_types: vec!["Patient".to_string()] };
        assert_eq!(reference.type_name(), None);
    }

    #[test]
    fn test_fhir_version_display() {
        assert_eq!(FhirVersion::R4.to_string(), "R4");
        assert_eq!(FhirVersion::R4B.to_string(), "R4B");
        assert_eq!(FhirVersion::R5.to_string(), "R5");
        assert_eq!(FhirVersion::R6.to_string(), "R6");
    }

    #[test]
    fn test_graph_metadata_default() {
        let metadata = GraphMetadata::default();
        assert_eq!(metadata.generator_version, env!("CARGO_PKG_VERSION"));
        assert!(metadata.source_packages.is_empty());
        assert!(!metadata.generated_at.is_empty());
    }
}
