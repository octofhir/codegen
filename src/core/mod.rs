//! Core functionality for codegen

pub mod error;
pub mod graph_builder;
pub mod ir;
pub mod parser;
pub mod resolver;

pub use error::{Error, Result};
pub use graph_builder::TypeGraphBuilder;
pub use ir::TypeGraph;
pub use parser::StructureDefinitionParser;
pub use resolver::SchemaResolver;
