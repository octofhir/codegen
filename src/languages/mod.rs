//! Language-specific code generation backends
//!
//! This module contains implementations for different target languages.

pub mod typescript;
