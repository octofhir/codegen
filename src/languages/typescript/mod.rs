//! TypeScript code generation

mod backend;
pub mod choice_generator;
pub mod datatype_generator;
/// Documentation generation for TypeScript with JSDoc and TypeDoc support
pub mod documentation_generator;
/// Helper methods generation for TypeScript
pub mod helpers_generator;
/// Package manifest generation for TypeScript
pub mod manifest_generator;
pub mod resource_generator;
/// Complete TypeScript SDK generation orchestrator
pub mod sdk_generator;
pub mod templates;
/// Validation functions generation for TypeScript
pub mod validation_generator;

pub use backend::TypeScriptBackend;
pub use choice_generator::ChoiceElementGenerator;
pub use datatype_generator::DatatypeGenerator;
pub use documentation_generator::DocumentationGenerator;
pub use helpers_generator::HelpersGenerator;
pub use manifest_generator::{ManifestGenerator, PackageConfig};
pub use resource_generator::ResourceGenerator;
pub use sdk_generator::TypeScriptSdkGenerator;
pub use validation_generator::ValidationGenerator;
