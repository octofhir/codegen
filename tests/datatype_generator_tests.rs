use insta::assert_snapshot;
use octofhir_codegen::core::ir::*;
use octofhir_codegen::languages::typescript::{DatatypeGenerator, TypeScriptBackend};

fn create_human_name() -> DataType {
    DataType {
        name: "HumanName".to_string(),
        base: Some("Element".to_string()),
        properties: vec![
            Property {
                name: "use".to_string(),
                path: "HumanName.use".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/name-use".to_string(),
                    description: Some("The use of a human name.".to_string()),
                }),
                constraints: vec![],
                short_description: "usual | official | temp | nickname | anonymous | old | maiden".to_string(),
                definition: "Identifies the purpose for this name".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "text".to_string(),
                path: "HumanName.text".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Text representation of the full name".to_string(),
                definition: "Specifies the entire name as it should be displayed".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "family".to_string(),
                path: "HumanName.family".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Family name (often called 'Surname')".to_string(),
                definition: "The part of a name that links to the genealogy".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "given".to_string(),
                path: "HumanName.given".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Given names (not always 'first'). Includes middle names".to_string(),
                definition: "Given name".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "prefix".to_string(),
                path: "HumanName.prefix".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Parts that come before the name".to_string(),
                definition: "Part of the name that is acquired as a title due to academic, legal, employment or nobility status".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "suffix".to_string(),
                path: "HumanName.suffix".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Parts that come after the name".to_string(),
                definition: "Part of the name that is acquired as a title due to academic, legal, employment or nobility status".to_string(),
                comments: None,
                examples: vec![],
            },
        ],
        documentation: Documentation {
            short: "Name of a human - parts and usage".to_string(),
            definition: "A human's name with the ability to identify parts and usage.".to_string(),
            url: Some("http://hl7.org/fhir/StructureDefinition/HumanName".to_string()),
            ..Default::default()
        },
        url: "http://hl7.org/fhir/StructureDefinition/HumanName".to_string(),
        is_abstract: false,
    }
}

fn create_address() -> DataType {
    DataType {
        name: "Address".to_string(),
        base: Some("Element".to_string()),
        properties: vec![
            Property {
                name: "use".to_string(),
                path: "Address.use".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: true,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/address-use".to_string(),
                    description: Some("The use of an address.".to_string()),
                }),
                constraints: vec![],
                short_description: "home | work | temp | old | billing".to_string(),
                definition: "The purpose of this address".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "type".to_string(),
                path: "Address.type".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/address-type".to_string(),
                    description: Some("The type of an address.".to_string()),
                }),
                constraints: vec![],
                short_description: "postal | physical | both".to_string(),
                definition: "Distinguishes between physical addresses and mailing addresses".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "line".to_string(),
                path: "Address.line".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional_array(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Street name, number, direction & P.O. Box etc.".to_string(),
                definition: "This component contains the house number, apartment number, street name, street direction, P.O. Box number, delivery hints, and similar address information.".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "city".to_string(),
                path: "Address.city".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Name of city, town etc.".to_string(),
                definition: "The name of the city, town, suburb, village or other community or delivery center".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "state".to_string(),
                path: "Address.state".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Sub-unit of country".to_string(),
                definition: "Sub-unit of a country with limited sovereignty in a federally organized country".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "postalCode".to_string(),
                path: "Address.postalCode".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Postal code for area".to_string(),
                definition: "A postal code designating a region defined by the postal service".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "country".to_string(),
                path: "Address.country".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Country (e.g. can be ISO 3166 2 or 3 letter code)".to_string(),
                definition: "Country - a nation as commonly understood or generally accepted".to_string(),
                comments: None,
                examples: vec![],
            },
        ],
        documentation: Documentation {
            short: "An address expressed using postal conventions (as opposed to GPS or other location definition formats)".to_string(),
            definition: "An address expressed using postal conventions (as opposed to GPS or other location definition formats). This data type may be used to convey addresses for use in delivering mail as well as for visiting locations which might not be valid for mail delivery. There are a variety of postal address formats defined around the world.".to_string(),
            url: Some("http://hl7.org/fhir/StructureDefinition/Address".to_string()),
            ..Default::default()
        },
        url: "http://hl7.org/fhir/StructureDefinition/Address".to_string(),
        is_abstract: false,
    }
}

fn create_quantity() -> DataType {
    DataType {
        name: "Quantity".to_string(),
        base: Some("Element".to_string()),
        properties: vec![
            Property {
                name: "value".to_string(),
                path: "Quantity.value".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "decimal".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Numerical value (with implicit precision)".to_string(),
                definition: "The value of the measured amount".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "comparator".to_string(),
                path: "Quantity.comparator".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: true,
                is_summary: true,
                binding: Some(ValueSetBinding {
                    strength: BindingStrength::Required,
                    value_set: "http://hl7.org/fhir/ValueSet/quantity-comparator".to_string(),
                    description: Some("How the Quantity should be understood and represented.".to_string()),
                }),
                constraints: vec![],
                short_description: "< | <= | >= | >".to_string(),
                definition: "How the value should be understood and represented".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "unit".to_string(),
                path: "Quantity.unit".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "string".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Unit representation".to_string(),
                definition: "A human-readable form of the unit".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "system".to_string(),
                path: "Quantity.system".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "uri".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "System that defines coded unit form".to_string(),
                definition: "The identification of the system that provides the coded form of the unit".to_string(),
                comments: None,
                examples: vec![],
            },
            Property {
                name: "code".to_string(),
                path: "Quantity.code".to_string(),
                property_type: PropertyType::Primitive {
                    type_name: "code".to_string(),
                },
                cardinality: CardinalityRange::optional(),
                is_choice: false,
                choice_types: vec![],
                is_modifier: false,
                is_summary: true,
                binding: None,
                constraints: vec![],
                short_description: "Coded form of the unit".to_string(),
                definition: "A computer processable form of the unit in some unit representation system".to_string(),
                comments: None,
                examples: vec![],
            },
        ],
        documentation: Documentation {
            short: "A measured or measurable amount".to_string(),
            definition: "A measured amount (or an amount that can potentially be measured). Note that measured amounts include amounts that are not precisely quantified, including amounts involving arbitrary units and floating currencies.".to_string(),
            url: Some("http://hl7.org/fhir/StructureDefinition/Quantity".to_string()),
            ..Default::default()
        },
        url: "http://hl7.org/fhir/StructureDefinition/Quantity".to_string(),
        is_abstract: false,
    }
}

#[test]
fn test_generate_human_name() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);
    let datatype = create_human_name();

    let output = generator.generate_datatype(&datatype).unwrap();
    assert_snapshot!("human_name_datatype", output);
}

#[test]
fn test_generate_address() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);
    let datatype = create_address();

    let output = generator.generate_datatype(&datatype).unwrap();
    assert_snapshot!("address_datatype", output);
}

#[test]
fn test_generate_quantity() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);
    let datatype = create_quantity();

    let output = generator.generate_datatype(&datatype).unwrap();
    assert_snapshot!("quantity_datatype", output);
}

#[test]
fn test_generate_primitive_wrapper() {
    let output = DatatypeGenerator::<TypeScriptBackend>::generate_primitive_wrapper().unwrap();
    assert_snapshot!("fhir_primitive_wrapper", output);
}

#[test]
fn test_generate_simple_primitive_types() {
    let output = DatatypeGenerator::<TypeScriptBackend>::generate_simple_primitive_types().unwrap();
    assert_snapshot!("simple_primitive_types", output);
}

#[test]
fn test_generate_primitive_types() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);

    let mut graph = TypeGraph::new(FhirVersion::R4);

    // Add some sample primitives
    graph.add_primitive(
        "boolean".to_string(),
        PrimitiveType {
            name: "boolean".to_string(),
            base: Some("Element".to_string()),
            pattern: None,
            documentation: Documentation {
                short: "true | false".to_string(),
                definition: "Value of \"true\" or \"false\"".to_string(),
                ..Default::default()
            },
            url: "http://hl7.org/fhir/StructureDefinition/boolean".to_string(),
        },
    );

    graph.add_primitive(
        "string".to_string(),
        PrimitiveType {
            name: "string".to_string(),
            base: Some("Element".to_string()),
            pattern: None,
            documentation: Documentation {
                short: "A sequence of Unicode characters".to_string(),
                definition: "A sequence of Unicode characters".to_string(),
                ..Default::default()
            },
            url: "http://hl7.org/fhir/StructureDefinition/string".to_string(),
        },
    );

    let output = generator.generate_primitive_types(&graph).unwrap();
    assert_snapshot!("primitive_types_with_wrapper", output);
}

#[test]
fn test_collect_human_name_dependencies() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);
    let datatype = create_human_name();

    let deps = generator.collect_dependencies(&datatype);

    assert!(deps.contains("Element"));
    assert!(!deps.contains("string")); // Primitives should not be included
    assert!(!deps.contains("code"));
}

#[test]
fn test_generate_datatype_file() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);
    let datatype = create_human_name();

    let deps = generator.collect_dependencies(&datatype);
    let output = generator.generate_datatype_file(&datatype, &deps).unwrap();

    assert_snapshot!("human_name_file_complete", output);
}

#[test]
fn test_generate_datatypes_index() {
    let backend = TypeScriptBackend::new();
    let generator = DatatypeGenerator::new(backend);

    let mut graph = TypeGraph::new(FhirVersion::R4);
    graph.add_datatype("HumanName".to_string(), create_human_name());
    graph.add_datatype("Address".to_string(), create_address());
    graph.add_datatype("Quantity".to_string(), create_quantity());

    let output = generator.generate_datatypes_index(&graph).unwrap();
    assert_snapshot!("datatypes_index", output);
}
