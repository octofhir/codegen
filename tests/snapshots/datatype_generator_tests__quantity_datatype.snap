---
source: tests/datatype_generator_tests.rs
expression: output
---
/**
 * A measured or measurable amount
 * 
 * A measured amount (or an amount that can potentially be measured). Note that
 * measured amounts include amounts that are not precisely quantified, including
 * amounts involving arbitrary units and floating currencies.
 * 
 * @see http://hl7.org/fhir/StructureDefinition/Quantity
 */
export interface Quantity extends Element {
  value?: number;
  comparator?: string;
  unit?: string;
  system?: string;
  code?: string;
}
